p2p/src/utils/bloom_filters/rolling_bloom_filter.rs

```diff
@@ -0,0 +1,105 @@
+/// Then that the oldest subfilter is replaced by a new one.
+/// An element is present if it's present in any subfilter.
+pub struct RollingBloomFilter<T> {
+    /// The list of subfilters that store items.
+    /// The oldest subfilter will be replaced by a new empty subfilter when they are full.
+    subfilters: [BloomFilter<T>; 3],
+
+    /// Currently used subfilter index
+    subfilter_index: usize,
+
+    /// Number of items added to the current subfilter
+    count_subfilter: usize,
+
+    /// Maximum number of items that can be added to a subfilter
+    size_subfilter: usize,
+
+    /// Desired false positive probability in each subfilter
+    fpp_subfilter: f64,
+}
+
+impl<T: Hash> RollingBloomFilter<T> {
+    /// Constructs a new rolling bloom filter.
+    ///
+    /// # Arguments
+    /// `size` - maximum number of items to store in the filter (up to 3/2*size of last items can be stored at times)
+    /// `fpp` - required false positive rate
+    /// `rng` - random number generator used to randomize hashes
+    pub fn new(size: usize, fpp: f64, rng: &mut impl Rng) -> Self {
+        assert!(size > 0);
+        assert!(fpp > 0.0 && fpp < 1.0);
+
+        // Use the smaller bloom filters to store size/2 of the last items in each.
+        // This way we always remember `size..3/2*size` items.
+        let size_per_filter = (size + 1) / 2;
+
+        // `fpp_per_filter` must be derived from `fpp` so that the probability of being detected in any subfilter is about `fpp`.
+        // Since each subfilter can be considered independent, it is about (1 - p)^3,
+        // which can be approximated by 1-3*p (if p is small).
+        // As a result, the required ffp per filter should be about fpp / 3.0.
+        let fpp_per_filter = fpp / 3.0;
+
+        RollingBloomFilter {
+            subfilters: [
+                BloomFilter::new(size_per_filter, fpp_per_filter, rng),
+                BloomFilter::new(size_per_filter, fpp_per_filter, rng),
+                BloomFilter::new(size_per_filter, fpp_per_filter, rng),
+            ],
+            subfilter_index: 0,
+            count_subfilter: 0,
+            size_subfilter: size_per_filter,
+            fpp_subfilter: fpp_per_filter,
+        }
+    }
+
+    /// Add item to the rolling bloom filter.
+    pub fn insert(&mut self, item: &T, rng: &mut impl Rng) {
+        self.subfilters[self.subfilter_index].insert(item);
+        self.count_subfilter += 1;
+        if self.count_subfilter == self.size_subfilter {
+            // The maximum number of items in the current subfilter has been reached.
+            // Create a new subfilter with new seeds to get new false positives.
+            self.subfilter_index = (self.subfilter_index + 1) % 3;
+            self.subfilters[self.subfilter_index] =
+                BloomFilter::new(self.size_subfilter, self.fpp_subfilter, rng);
+            self.count_subfilter = 0;
```

Comment:
If count_subfilter is common among all bloom filters, and the filters run in a cyclic way... what's the difference between having a single filter with 3x the size and having 3 filters like this?

In fact, if size_subfilter is multiple of 3, then only 1 subfilter will be constantly reset and the others will be ignored and never reset. Am I missing something?